content/lib/2021-04-02-operating-system-interface.md

```diff
@@ -0,0 +1,225 @@
+    - EX: Do we need POSIX compliant user/group management? Or will Linux capabilities suffice?
+ - Management of required `node` level services can be safely controlled from `kubernetesland`
+    - EX: CNI mutations
+    - EX: Privileged DaemonSets for mutating a host
+    - EX: Kernel security controls (Seccomp, SELinux, eBPF, Auth, User Management)
+    - EX: Storage providers interacting with the host
+    
+# Prior Art
+
+It is obvious that there is prior art in the following `node` level services.
+There are fundamental flaws, and voids in what the existing interfaces assume.
+
+ ### Network (CNI)
+
+Official [Container Networking Interface](https://github.com/containernetworking/cni) standard.
+
+ - CNI defines the container networking API surface/shape for interacting with networking plugins [via the container runtime](https://github.com/containernetworking/cni/blob/master/SPEC.md#section-2-execution-protocol) in `userland`.
+     - CNI does not solve *how* to install and mutate a CNI service in `userland` from a `kubernetesland` perspective.
+         - CNI is [required by a container runtime](https://github.com/containernetworking/cni/blob/master/SPEC.md#overview-1), however it is assumed this is installed from `kubernetesland`. This is a paradox. 
+     - CNI assumes access to `userland` and more importantly access (and permission) to `fork(2)` and `execve(2)`
+     - CNI does not solve *how* access is managed, or acquired. 
+ - CNI does not offer a standardized way of reporting, measuring, or limiting resources in `kubernetesland`.
+ 
+### Storage (CSI)
+
+Official [Container Storage Interface](https://github.com/container-storage-interface/spec) standard.
+
+ - CSI defines the container storage API surface/shape for interfacing with block and file storage drivers in `userland`. 
+     - More information on [CSI drivers](https://kubernetes.io/blog/2019/01/15/container-storage-interface-ga/)
+     - More information on [Kubelet device registration and plugins](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/)
+ - CSI does not solve *how* to install or manage block or file storage drivers.
+ - The Kubelet plugin system does NOT solve how to install or manage Kubelet plugins from Kubernetes. 
+     - [Access to the Kubelet API server is required](https://v1-18.docs.kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-registration) to interact with the exposed registration service. 
+ - CSI (via the Kubelet registration endpoint and logic) can set limitations and provide status updates in `kubernetesland`.
+
+
+### Runtime (Compute) (CRI/OCI)
+
+Official [Container Runtime Interface](https://github.com/kubernetes/cri-api) API.
+Official [Open Container Interface](https://github.com/opencontainers/runtime-spec) standard.
+    - Official [Open Container Image Spec](https://github.com/opencontainers/image-spec)
+
+CRI is a implementation of OCI.
```

Comment:
@kris-nova CRI is not an implementation of OCI. CRI is just an interface that the container runtime must adopt to be able to use it with Kubernetes.

OCI is a spec used for the low-level container runtimes (like runc or crun). You can't hook the kubelet directly to runc, CRI defines more high-level stuff than OCI (like list pods, etc.). You hook the kubelet to a "high level" container runtime, like docker or containerd (that internally use runc to actually create containers). Runc is a low level container runtime that implements the OCI runtime spec (i.e. it can create containers out of files in the OCI runtime spec format).

The high level idea, IIUC, is like this: The kubelet speaks CRI with the high-level container runtime. The high-level container runtime is big thing (it can pull from registries, authenticate with registries, etc.) and after downloading the container image and all, creates a file using the OCI runtime-spec format (usually called `config.json`) that will be passed to the the low-level container runtime (like runc) to create the containers.

The flow is something like this: kubelet <--- CRI -- > Containerd <-- OCI runtime spec file ---> runc --> process.

In case I'm not clear with my ASCII art, I mean: The kubelet speaks CRI with containerd. Containerd has a plugin (now built-in) to speak CRI with the kubelet. When containerd needs to create a container, for example kubelet signaled we need a new pod, containerd creates a file that is in the OCI runtime spec format according to the container it needs to create, and then calls runc to actually create the container. Containerd really just uses runc, [it vendors runc][link].

Sorry for being late, I'm commenting here as you mentioned to comment on these. Of course, this doesn't change the issues you enumerate later :)


[link]: https://github.com/containerd/containerd/blob/ff05d499e29d3b7b88e88e91e9a80551d3c167b9/go.mod#L42